README.md
<h1 align="center">vtc-ex</h1>
<p align="center">
    <img height=150 class="heightSet" align="center" src="https://raw.githubusercontent.com/opencinemac/vtc-py/master/zdocs/source/_static/logo1.svg"/>
</p>
<p align="center">A SMPTE Timecode Library for Elixir</p>
<p align="center">
    <a href="https://dev.azure.com/peake100/Open%20Cinema%20Collective/_build?definitionId=19"><img src="https://dev.azure.com/peake100/Open%20Cinema%20Collective/_apis/build/status/vtc-ex?repoName=opencinemac%2Fvtc-ex&branchName=dev" alt="click to see build pipeline"></a>
    <a href="https://dev.azure.com/peake100/Open%20Cinema%20Collective/_build?definitionId=19"><img src="https://img.shields.io/azure-devops/tests/peake100/Open%20Cinema%20Collective/19/dev?compact_message" alt="click to see build pipeline"></a>
    <a href="https://dev.azure.com/peake100/Open%20Cinema%20Collective/_build?definitionId=19"><img src="https://img.shields.io/azure-devops/coverage/peake100/Open%20Cinema%20Collective/19/dev?compact_message" alt="click to see build pipeline"></a>
</p>
<p align="center">
    <a href="https://hex.pm/packages/vtc"><img src="https://img.shields.io/hexpm/v/vtc.svg" alt="PyPI version" height="18"></a>
    <a href="https://hexdocs.pm/vtc/readme.html"><img src="https://img.shields.io/badge/docs-hexdocs.pm-blue" alt="Documentation"></a>
</p>

## Demo

A small preview of what `Vtc` has to offer. Note that printing statements like 
`inspect/1` have been elided from the examples below.

`Vtc` represents specific frames in a video stream as [Framestamps](`Vtc.Framestamp`), 
which can [parsed](Vtc.Framestamp.html#parse) from a number of different formats, 
including SMPTE timecode, frame count, and physical film length measured in 
feet and frames:

```elixir
iex> Framestamp.with_seconds!(1.5, Rates.f23_98())
"<00:05:23:04 <23.98 NTSC>>"
iex> stamp = Framestamp.with_frames!("17:23:13:02", Rates.f23_98())
"<17:23:00:02 <23.98 NTSC>>"
```

Once in a [Framestamp](`Vtc.Framestamp`) struct, you 
[convert](Vtc.Framestamp.html#convert) to any of the supported formats:

```elixir
iex> Framestamp.smpte_timecode(stamp)
"00:05:23:04"
iex> Framestamp.frames(stamp)
1501922
iex> Framestamp.feet_and_frames(stamp)
"<93889+10 :ff35mm_4perf>"
```

[Comparisons](Vtc.Framestamp.html#compare) and 
[kernel sorting](Vtc.Framestamp.html#module-sorting-support) are supported, with many 
helper functions for specific comparisons:

```elixir
iex> Framestamp.compare(stamp, "02:00:00:00")
:gt
iex> Framestamp.gt?(stamp, "02:00:00:00")
true
iex> stamp_01 = Framestamp.with_frames!("01:00:00:00", Rates.f23_98())
iex> stamp_02 = Framestamp.with_frames!("02:00:00:00", Rates.f23_98())
iex> data_01 = %{id: 2, tc: stamp_01}
iex> data_02 = %{id: 1, tc: stamp_02}
iex> [data_02, data_01] |> Enum.sort_by(& &1.tc, Framestamp) |> inspect()
"[%{id: 2, tc: <01:00:00:00 <23.98 NTSC>>}, %{id: 1, tc: <02:00:00:00 <23.98 NTSC>>}]"
```

All sensible [arithmetic](Vtc.Framestamp.html#arithmetic) operations are provided, such 
as addition, subtraction, and multiplication:

```elixir
iex> stamp = Framestamp.add(tc, "01:00:00:00")
"<18:23:13:02 <23.98 NTSC>>"
```

You can even use native operators within special [eval/2](Vtc.Framestamp.html#eval/2) 
blocks:

```elixir
iex> Framestamp.eval at: 23.98 do
iex>   stamp + "00:30:00:00" * 2 - "00:15:00:00"
iex> end
"<19:08:13:02 <23.98 NTSC>>"
```

[Ranges](`Vtc.Framestamp.Range`) let you operate on in/out points, for instance, finding the 
overlapping area between two ranges:

```elixir
iex> a_in = Framestamp.with_frames!("01:00:00:00", Rates.f23_98())
iex> a = Framestamp.Range.new!(a_in, "02:00:00:00")
"<01:00:00:00 - 02:00:00:00 :exclusive <23.98 NTSC>>"
iex> b_in = Framestamp.with_frames!("01:45:00:00", Rates.f23_98())
iex> b = Framestamp.Range.new!(b_in, "02:30:00:00")
"<01:45:00:00 - 02:30:00:00 :exclusive <23.98 NTSC>>"
iex> Framestamp.Range.intersection!(a, b)
"<01:45:00:00 - 02:00:00:00 :exclusive <23.98 NTSC>>"
```

## Ecto types

Vtx ships with optional ecto types that can be used to accelerate you timecode workflow
at the database level:

```elixir
## Migration file

defmodule MyApp.MySchema do
  @moduledoc false
  use Ecto.Migration

  alias Vtc.Ecto.Postgres.PgFramestamp
  alias Vtc.Framestamp

  def change do
    create table("my_table", primary_key: false) do
      add(:id, :uuid, primary_key: true, null: false)
      add(:a, Framestamp.type())
      add(:b, Framestamp.type())
    end
  end
end
```

```elixir
## Schema file

defmodule Vtc.Test.Support.FramestampSchema01 do
  @moduledoc false
  use Ecto.Schema

  alias Ecto.Changeset
  alias Vtc.Framestamp

  @type t() :: %__MODULE__{
          id: Ecto.UUID.t(),
          stamp_in: Framestamp.t(),
          stamp_out: Framestamp.t()
        }

  @primary_key {:id, Ecto.UUID, autogenerate: true}

  schema "my_table" do
    field(:start, Framestamp)
    field(:end, Framestamp)
  end

  @spec changeset(%__MODULE__{}, %{atom() => any()}) :: Changeset.t(%__MODULE__{})
  def changeset(schema, attrs), do: Changeset.cast(schema, attrs, [:id, :stamp_in, :stamp_out])
end
```

Values can be used in Ecto queries using the [type/2](`Ecto.Query.API.type/2`) function.
Vtc registers native operators for each type, so you can write queries like you would
expect to:

```elixir
iex> one_hour = Framestamp.with_frames("01:00:00:00", Rates.f23_98())
iex> 
iex> EdlEvents
iex> |> where([event], event.stamp_in > type(^one_hour, Framestamp))
iex> |> select([event], {event, event.stamp_out - event.stamp_in})
```

The above query finds all events with a start time greater than `01:00:00:00` and
returns the record AND its calculated duration.

## Further Reading

Check out the [Quickstart Guide](quickstart.html) for a walkthrough of what `Vtc` 
offers for application-level code, and the [Ecto Quickstart Guide](ecto_quickstart.cheatmd)
for a deep-dive on working with Vtc's Postgres offerings.

The [API Reference](api-reference.html) offers a complete technical accounting of Vtc's 
capabilities.

## Goals

- Offer a comprehensive set of tools for parsing, manipulating and rendering timecode
  with all it's quirks and incarnations.

- Define an intuitive, idiomatic Elixir API.

- Do all calculations in Rational representation, so there is both no drift or rounding 
  errors when manipulating NTSC timecode, and we can represent time as finely as 
  possible rather than being limited to the granularity of frame numbers.

- Be approachable for newcomers to the timecode problem space. Each function and concept
  in the [API Reference](api-reference.html) includes a primer on what it is and where 
  it is used in Film and Television workflow.

- Offer a flexible set of tools that support both rigorous, production-quality code and
  quick scratch scripts.

## Features

- SMPTE Conventions:
    - [X] NTSC
    - [X] Drop-Frame
    - [ ] Interlaced timecode
- Timecode Representations:
    - [X] Framestamp  | 18018/5 seconds @ 24000/1001
    - [X] Timecode    | '01:00:00:00'
    - [X] Frames      | 86400
    - [X] Seconds     | 3600.0
    - [X] Runtime     | '01:00:00.0'
    - [X] Rational    | 18018/5
    - [X] Feet+Frames | '5400+00'
        - [X] 35mm, 4-perf
        - [ ] 35mm, 3-perf
        - [X] 35mm, 2-perf
        - [X] 16mm
    - [X] Premiere Ticks | 15240960000000
- Operations:
    - [X] Comparisons (==, <, <=, >, >=)
    - [X] Add
    - [X] Subtract
    - [X] Scale (multiply and divide)
    - [X] Divmod
    - [X] Modulo
    - [X] Negative
    - [X] Absolute
    - [X] Rebase (recalculate frame count at new framerate)
    - [X] Native Operator Evaluation
- Flexible Parsing:
    - [X] Partial timecodes      | '1:12'
    - [X] Partial runtimes       | '1.5'
    - [X] Negative string values | '-1:12', '-3+00'
    - [X] Poorly formatted SMPTE timecode    | '1:13:4'
- [X] Built-in consts for common framerates.
- [X] Configurable rounding options.
- [X] Support for standard library sorting behavior.
- [X] Range type for working with and comparing frame ranges.
    - [X] Overlap between ranges
    - [X] Distance between ranges
    - [X] Inclusive and exclusive ranges
- [X] Postgres Composite Types with Ecto and Postgrex:
    - Rational[X]
      - [X] Native comparison operators
      - [X] Native arithmetic operators
      - [X] Native BTree index support
    - Framerate[X]
      - [X] Native comparison operators
    - Framestamp[X]
      - [X] Native comparison operators
      - [X] Native arithmetic operators
      - [X] Native BTree index support
      - [X] Native inspection functions
    - Framestamp.Range
      - [X] Native GiST index support
      - [X] Native inspection functions

## Attributions

Drop-frame calculations adapted from 
[David Heidelberger's blog](https://www.davidheidelberger.com/2010/06/10/drop-frame-timecode/).

35mm, 2perf and 16mm format support based on 
[Jamie Hardt's work for vtc-rs](https://github.com/opencinemac/vtc-rs/pull/8).

Logo made by Freepik from [www.flaticon.com](https://www.flaticon.com/).

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `vtc` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:vtc, "~> 0.14"}
  ]
end
```

zdocs/history.md
# Timecode: A History

But first: what is timecode?

If you're already familiar with timecode, it's history, and it's flavors, feel free to
skip this section.

## How we got here

Back in the days of film, a running strip of numbers ran along the edge of the film
stock to uniquely identify each frame, called
[keycode](https://en.wikipedia.org/wiki/Keykode)

Keycode was essential to the film editing process. The raw negative of a film is
irreplaceable: you loose quality each time you make a copy. Editing film is necessarily
a [destructive process](https://nofilmschool.com/2017/06/editing-on-a-flatbed), and
often required multiple iterations. It would be just a tad nerve-wracking to take a pair
of scissors and some glue to the one-of-a-kind film reels straight out of the camera
on set, then running it over and over through a flatbed.

To avoid potential disaster, editors made their cut of the film using copies of the
raw negative, called a [work print](https://en.wikipedia.org/wiki/Workprint), allowing
the editor to work without fear of sinking a project from slicing, dicing, and wearing
at the film.

When the edit was complete, it was necessary to know *exactly* where the edits had been
made, so it could be recreated with the raw negative for finishing. A *cut list* would
be written out, with the exact reels and keycodes for every cut, and would be used to
make an exact duplicate of the editor's work print with the mint condition raw negative.

In video and digital filmmaking, the same approach is used. Massive RAW files from a
RED, ARRI, Sony, or other cinema camera are rendered down to more manageable files an
Editor's machine won't choke on. Once the edit is complete, the raw files are
re-assembled using a digital cutlist on a powerful machine for finishing out the film.

In film, we referenced *keycode* to know exactly what frame was being displayed on
screen at any given time. In digital video, we reference the *timecode* of a given
frame, as defined by the SMPTE Timecode standard.

## Further Reading

For a technical deep-dive into the many flavors of timecode, check out
[Frame.io's](frame.io)
[excellent blog post](https://blog.frame.io/2017/07/17/timecode-and-frame-rates) on
the subject.

lib/sources.ex
defmodule Vtc.Source do
  @moduledoc """
  Protocols for source values that can be used to construct a timecode.
  """
  alias Vtc.Timecode
  alias Vtc.Utils.Rational

  defprotocol Seconds do
    @moduledoc """
    Protocol which types can implement to be passed as the main value of
    `Timecode.with_seconds/2`.

    ## Implementations

    Out of the box, this protocol is implemented for the following types:

    - `Ratio`
    - `Integer`
    - `Float`
    - `String`
      - runtime ("01:00:00.0")
      - decimal ("3600.0")
    """

    alias Vtc.Framerate
    alias Vtc.Source
    alias Vtc.Timecode

    @typedoc """
    Result type of `Source.Seconds.seconds/2`.
    """
    @type result() :: {:ok, Rational.t()} | {:error, Timecode.ParseError.t()}

    @doc """
    Returns the value as a rational seconds value.

    ## Arguments

    - **value**: The source value.

    - **rate**: The framerate of the timecode being parsed.

    ## Returns

    A result tuple with a rational representation of the seconds value using `Ratio` on
    success.
    """
    @spec seconds(t(), Framerate.t()) :: result()
    def seconds(value, rate)
  end

  defimpl Seconds, for: [Ratio, Integer] do
    alias Vtc.Framerate
    alias Vtc.Private.Parse
    alias Vtc.Source
    alias Vtc.Utils.Rational

    @spec seconds(Rational.t(), Framerate.t()) :: Seconds.result()
    def seconds(value, rate), do: Parse.from_seconds_core(value, rate)
  end

  defimpl Seconds, for: Float do
    alias Vtc.Framerate
    alias Vtc.Source

    @spec seconds(float(), Framerate.t()) :: Seconds.result()
    def seconds(value, rate), do: value |> Ratio.new(1) |> Seconds.seconds(rate)
  end

  defimpl Seconds, for: [String, BitString] do
    alias Vtc.Framerate
    alias Vtc.Private.Parse
    alias Vtc.Source

    @spec seconds(String.t(), Framerate.t()) :: Seconds.result()
    def seconds(value, rate), do: Parse.parse_runtime_string(value, rate)
  end

  defprotocol Frames do
    @moduledoc """
    Protocol which types can implement to be passed as the main value of
    `Timecode.with_frames/2`.

    ## Implementations

    Out of the box, this protocol is implemented for the following types:

    - `Integer`
    - `String` & 'BitString'
      - timecode ("01:00:00:00")
      - integer ("86400")
      - Feet+Frames ("5400+00")
    """

    alias Vtc.Framerate
    alias Vtc.Source
    alias Vtc.Timecode

    @typedoc """
    Result type of `Vtc.Source.Frames.frames/2 aaa`.
    """
    @type result() :: {:ok, integer()} | {:error, Timecode.ParseError.t()}

    @doc """
    Returns the value as a frame count.

    # Arguments

    - **value**: The source value.

    - **rate**: The framerate of the timecode being parsed.

    # Returns

    A result tuple with an integer value representing the frame count on success.
    """

    @spec frames(t(), Framerate.t()) :: result()
    def frames(value, rate)
  end

  defimpl Frames, for: Integer do
    alias Vtc.Framerate
    alias Vtc.Source.Frames

    @spec frames(integer(), Framerate.t()) :: Frames.result()
    def frames(value, _rate), do: {:ok, value}
  end

  defimpl Frames, for: [String, BitString] do
    alias Vtc.Framerate
    alias Vtc.Private.Parse
    alias Vtc.Source.Frames

    @spec frames(String.t(), Framerate.t()) :: Frames.result()
    def frames(value, rate), do: Parse.parse_frames_string(value, rate)
  end

  defprotocol PremiereTicks do
    @moduledoc """
    Protocol which types can implement to be passed as the main value of
    `Vtc.Timecode.with_premiere_ticks/2`.

    # Implementations

    Out of the box, this protocol is implemented for the following types:

    - `Integer`
    """

    @doc """
    Returns the number of Adobe Premiere Pro ticks as an integer.

    # Arguments

    - **value**: The source value.

    - **rate**: The framerate of the timecode being parsed.

    # Returns

    A result tuple with a rational representation of the seconds value using `Ratio` on
    success.
    """

    alias Vtc.Framerate
    alias Vtc.Source
    alias Vtc.Timecode

    @typedoc """
    Result type of `ticks/2`.
    """
    @type result() :: {:ok, integer()} | {:error, Timecode.ParseError.t()}

    @spec ticks(t(), Framerate.t()) :: result()
    def ticks(value, rate)
  end

  defimpl PremiereTicks, for: Integer do
    alias Vtc.Framerate
    alias Vtc.Source

    @spec ticks(integer(), Framerate.t()) :: PremiereTicks.result()
    def ticks(value, _rate), do: {:ok, value}
  end
end

defmodule Vtc.Private.Parse do
  @moduledoc false

  use Ratio

  alias Vtc.Framerate
  alias Vtc.Private.Consts
  alias Vtc.Private.DropFrame
  alias Vtc.Source
  alias Vtc.Source.Frames
  alias Vtc.Source.Seconds
  alias Vtc.Timecode
  alias Vtc.Utils.Rational

  @spec from_seconds_core(Rational.t(), Framerate.t()) :: Seconds.result()
  def from_seconds_core(input, rate) do
    # If the vaue doesn't cleany divide into the framerate then we need to round to the
    # nearest frame.
    case Ratio.div(input, rate.playback) do
      %Ratio{} ->
        frames = rate.playback |> Ratio.mult(input) |> Rational.round()
        {:ok, Ratio.div(frames, rate.playback)}

      _ ->
        {:ok, input}
    end
  end

  @spec parse_frames_string(String.t(), Framerate.t()) :: Frames.result()
  def parse_frames_string(value, rate) do
    case parse_tc_string(value, rate) do
      {:ok, _} = result -> result
      {:error, %Timecode.ParseError{reason: :bad_drop_frames}} = error -> error
      {:error, _} -> parse_feet_and_frames(value, rate)
    end
  end

  @tc_regex ~r/^(?P<negative>-)?((?P<section_1>[0-9]+)[:|;])?((?P<section_2>[0-9]+)[:|;])?((?P<section_3>[0-9]+)[:|;])?(?P<frames>[0-9]+)$/

  @spec parse_tc_string(String.t(), Framerate.t()) :: Frames.result()
  def parse_tc_string(value, rate) do
    with {:ok, matched} <- apply_regex(@tc_regex, value) do
      matched
      |> tc_matched_to_sections()
      |> tc_sections_to_frames(rate)
    end
  end

  @spec apply_regex(Regex.t(), String.t()) :: {:ok, map()} | {:error, Timecode.ParseError.t()}
  defp apply_regex(regex, value) do
    regex
    |> Regex.named_captures(value)
    |> then(fn
      matched when is_map(matched) -> {:ok, matched}
      nil -> {:error, %Timecode.ParseError{reason: :unrecognized_format}}
    end)
  end

  # Extract TC sections from regex match.
  @spec tc_matched_to_sections(map()) :: Timecode.Sections.t()
  defp tc_matched_to_sections(matched) do
    negative? = Map.fetch!(matched, "negative") == "-"

    sections = extract_time_sections(matched, 3)

    {seconds, sections} = pop_time_section(sections)
    {minutes, sections} = pop_time_section(sections)
    {hours, _} = pop_time_section(sections)
    frames = matched |> Map.fetch!("frames") |> String.to_integer()

    %Timecode.Sections{
      negative?: negative?,
      hours: hours,
      minutes: minutes,
      seconds: seconds,
      frames: frames
    }
  end

  # Extracts a set of sections in a time string of format xx:yy:.. that may or may not
  # be truncated at the head.
  #
  # The regex matches are expected to have a series of fields like "section_1",
  # "section_2", etc that denote present sections whose meaning depends on the  number
  # of sections present.
  @spec extract_time_sections(map(), non_neg_integer()) :: [String.t()]
  defp extract_time_sections(regex_matches, section_count) do
    1..section_count
    |> Enum.map(&Integer.to_string/1)
    |> Enum.reduce([], fn section_index, sections ->
      case Map.fetch!(regex_matches, "section_#{section_index}") do
        "" -> sections
        this_section -> [this_section | sections]
      end
    end)
  end

  # Pops the next section at the end of the list and returns it as an integer.
  #
  # Returns `0` if the value is not present
  @spec pop_time_section([String.t()]) :: {integer(), [String.t()]}
  defp pop_time_section(["" | remaining]), do: {0, remaining}
  defp pop_time_section([value | remaining]), do: {String.to_integer(value), remaining}
  defp pop_time_section([]), do: {0, []}

  # Converts all TC fields to a total frame count
  @spec tc_sections_to_frames(Timecode.Sections.t(), Framerate.t()) :: Frames.result()
  defp tc_sections_to_frames(sections, rate) do
    with {:ok, adjustment} <- DropFrame.parse_adjustment(sections, rate) do
      frames_per_second = Framerate.timebase(rate)

      sections.seconds
      |> Ratio.add(sections.minutes * Consts.seconds_per_minute())
      |> Ratio.add(sections.hours * Consts.seconds_per_hour())
      |> Ratio.mult(frames_per_second)
      |> Ratio.add(sections.frames)
      |> Ratio.add(adjustment)
      |> Rational.round()
      |> then(fn frames -> if sections.negative?, do: -frames, else: frames end)
      |> then(&{:ok, &1})
    end
  end

  @ff_regex ~r/(?P<negative>-)?(?P<feet>[0-9]+)\+(?P<frames>[0-9]+)/

  @spec parse_feet_and_frames(String.t(), Framerate.t()) :: Frames.result()
  defp parse_feet_and_frames(value, rate) do
    with {:ok, groups} <- apply_regex(@ff_regex, value) do
      negative? = Map.fetch!(groups, "negative") == "-"
      feet = groups |> Map.fetch!("feet") |> String.to_integer()

      groups
      |> Map.fetch!("frames")
      |> String.to_integer()
      |> Ratio.add(feet * Consts.frames_per_foot())
      |> then(fn frames -> if negative?, do: -frames, else: frames end)
      |> Frames.frames(rate)
    end
  end

  @runtime_regex ~r/^(?P<negative>-)?((?P<section_1>[0-9]+)[:|;])?((?P<section_2>[0-9]+)[:|;])?(?P<seconds>[0-9]+(\.[0-9]+)?)$/

  @spec parse_runtime_string(String.t(), Framerate.t()) :: Seconds.result()
  def parse_runtime_string(value, rate) do
    with {:ok, matched} <- apply_regex(@runtime_regex, value) do
      matched
      |> runtime_matched_to_second()
      |> Source.Seconds.seconds(rate)
    end
  end

  @spec runtime_matched_to_second(map()) :: Rational.t()
  defp runtime_matched_to_second(matched) do
    negative? = Map.fetch!(matched, "negative") == "-"
    sections = extract_time_sections(matched, 2)

    {minutes, sections} = pop_time_section(sections)
    {hours, _} = pop_time_section(sections)

    matched
    |> Map.fetch!("seconds")
    |> Decimal.new()
    |> Ratio.new(1)
    |> Ratio.add(hours * Consts.seconds_per_hour())
    |> Ratio.add(minutes * Consts.seconds_per_minute())
    |> then(fn seconds -> if negative?, do: -seconds, else: seconds end)
  end
end
